src/librustc_typeck/check/regionck.rs

```diff
@@ -1395,74 +1484,330 @@ fn link_reborrowed_region<'a, 'tcx>(rcx: &Rcx<'a, 'tcx>,
+    // inference.  This is true because the region bounds we can
+    // derive for `T` never involve region variables -- they are
+    // always free regions.  The only place a region variable can come
+    // is on the RHS, and in that case, the smaller the region, the
+    // better. This means that our current inference, which always
+    // infers the smallest region it can, can just be used, and we'll
+    // know what the smallest value for `'r` is when it's done. We can
+    // then compare that to the regions in the LHS, which are already
+    // as big as possible, and we're all done.
+    //
+    // Projections can in fact be this simple as well. In particular,
+    // if the parameters `P0..Pn` do not involve any region variables,
+    // that's the same situation.
+    //
+    // Where things get thorny is when region variables are involved,
+    // because in that case relating `Pi: 'r` may influence the
+    // inference process, since it could cause `'r` to be inferred to
+    // a larger value. But the problem is that if we add that as a
+    // constraint into our dataflow graph, we've essentially committed
+    // to using option 1 (above) to show that `<P0 as
+    // Trait<P1..Pn>>::Foo: 'r` is satisfied, and it may be that
+    // Option 1 does not apply, but Option 2 or 3 does. But we can't
+    // know that now.
+    //
+    // For now we choose to accept this. It's a conservative choice,
+    // so we can move to a more sophisticated inference model later.
+    // And it's sometimes possible to workaround by introducing
+    // explicit type parameters or type annotations. But it ain't
+    // great!
+
+    let declared_bounds = projection_declared_bounds(rcx, origin.span(), projection_ty);
+
+    debug!("projection_must_outlive: declared_bounds={:?}",
+           declared_bounds);
+
+    // If we know that the projection outlives 'static, then we're done here.
+    if declared_bounds.contains(&ty::ReStatic) {
+        return;
+    }
 
-    // In the case of a projection T::Foo, we may be able to extract bounds from the trait def:
-    match *generic {
-        GenericKind::Param(..) => { }
-        GenericKind::Projection(ref projection_ty) => {
-            param_bounds.push_all(
-                &projection_bounds(rcx, origin.span(), projection_ty));
+    // Determine whether any of regions that appear in the projection
+    // were declared as bounds by the user. This is typically a situation
+    // like this:
+    //
+    //     trait Foo<'a> {
+    //         type Bar: 'a;
+    //     }
+    //
+    // where we are checking `<T as Foo<'_#0r>>: '_#1r`. In such a
+    // case, if we use the conservative rule, we will check that
+    // BOTH of the following hold:
+    //
+    //     T: _#1r
+    //     _#0r: _#1r
+    //
+    // This is overkill, since the declared bounds tell us that the
+    // the latter is sufficient.
+    let intersection_bounds: Vec<_> =
```

Comment:
Need moar comments :-)

`declared_bounds` is essentially the outlives bounds from predicates. `ProjectionCandidate` outlive-predicates only contain types that come from the trait-ref, and where-clause predicates can't really contain infer-variables (except for instantiations of higher-rank predicates, e.g. `for<'b> <Foo<'a, 'b> as Iterator>::Item: 'b`, but these ~~are still a subset of the trait-ref~~don't do anything useful - the relevant code at `ty::required_region_bounds` is *really* suspicious), so the intersection here seems strange.

Also, if there are *any* explicit bounds, the verify path would be less conservative than the `mk_subr` path.

I would prefer to:
* If there are no predicate bounds, require all components to outlive.
* If there is exactly 1 unique predicate bound, and it is an inference variable, require it to outlive. This should mainly occur in the `trait Foo<'a> { type Ty: 'a; }` case.
* Otherwise, verify.

Having multiple predicate bounds isn't exactly trivial (you need to e.g. `trait Foo<'a, 'b> { type Assoc: 'a+'b; }`), and requiring *both* parameters to live long enough is kind-of counterintuitive.